testcases/cve/stack_clash.c

```diff
@@ -0,0 +1,255 @@
+		"mmap = [%lx, %lx), addr = %lx, diff = %lx, THRESHOLD = %lx",
+		mapped_addr, mmap_end, fault_addr, diff, THRESHOLD);
+	if (diff < 0 || (unsigned long)diff < THRESHOLD)
+		_exit(EXIT_FAILURE);
+	else
+		_exit(EXIT_SUCCESS);
+}
+
+unsigned long read_stack_addr_from_proc(unsigned long *stack_size)
+{
+	FILE *fh;
+	char buf[1024];
+	unsigned long stack_top = -1UL, start, end;
+
+	fh = SAFE_FOPEN("/proc/self/maps", "r");
+
+	while (!feof(fh)) {
+		if (fgets(buf, sizeof(buf), fh) == NULL) {
+			tst_brk(TBROK | TERRNO, "fgets");
+			goto out;
+		}
+
+		if (!strstr(buf, "[stack"))
+			continue;
+
+		if (sscanf(buf, "%lx-%lx", &start, &end) != 2) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			goto out;
+		}
+
+		*stack_size = end - start;
+
+		if (STACK_GROWSDOWN)
+			stack_top = start;
+		else
+			stack_top = end;
+		break;
+	}
+
+out:
+	SAFE_FCLOSE(fh);
+	return stack_top;
+}
+
+void do_child(void)
+{
+	unsigned long stack_addr, stack_size;
+	stack_t signal_stack;
+	struct sigaction segv_sig = {.sa_sigaction = segv_handler, .sa_flags = SA_ONSTACK|SA_SIGINFO};
+	void *map;
+	unsigned long gap = GAP_PAGES * PAGE_SIZE;
+	struct rlimit rlimit;
+
+	stack_addr = read_stack_addr_from_proc(&stack_size);
+	if (stack_addr == -1UL) {
+		tst_brk(TBROK, "can't read stack top from /proc/self/maps");
+		return;
+	}
+
+	SAFE_GETRLIMIT(RLIMIT_STACK, &rlimit);
+	if (rlimit.rlim_cur != RLIM_INFINITY && stack_size >= rlimit.rlim_cur) {
+		rlimit.rlim_cur = stack_size + PAGE_SIZE;
+		rlimit.rlim_max = rlimit.rlim_cur;
+		SAFE_SETRLIMIT(RLIMIT_STACK, &rlimit);
+	}
+
+	if (STACK_GROWSDOWN)
+		mapped_addr = stack_addr - gap - MAPPED_LEN;
+	else
+		mapped_addr = stack_addr + gap;
+
+	mapped_addr &= PAGE_MASK;
+	map = SAFE_MMAP((void *)mapped_addr, MAPPED_LEN,
+			PROT_READ|PROT_WRITE,
+			MAP_ANON|MAP_PRIVATE|MAP_FIXED, -1, 0);
+	tst_res(TINFO, "Stack:0x%lx+0x%lx mmap:%p+0x%lx",
+		stack_addr, stack_size, map, MAPPED_LEN);
+
+	signal_stack.ss_sp = SAFE_MALLOC(SIGNAL_STACK_SIZE);
+	signal_stack.ss_size = SIGNAL_STACK_SIZE;
+	signal_stack.ss_flags = 0;
+	if (sigaltstack(&signal_stack, NULL) == -1) {
+		tst_brk(TBROK | TERRNO, "sigaltstack");
+		return;
+	}
+	if (sigaction(SIGSEGV, &segv_sig, NULL) == -1 ||
+	    sigaction(SIGBUS,  &segv_sig, NULL) == -1) {
+		tst_brk(TBROK | TERRNO, "sigaction");
+		return;
+	}
+
+#ifdef DEBUG
+	{
+		char buf[128];
+		const char *cmd[] = {"cat", buf, NULL};
+		sprintf(buf, "/proc/%d/maps", getpid());
+		tst_run_cmd(cmd, NULL, NULL, 0);
+	}
+#endif
+
+	exhaust_stack_into_sigsegv();
+}
+
+void setup(void)
+{
+	char buf[4096], *p;
+
+	PAGE_SIZE = sysconf(_SC_PAGESIZE);
+	PAGE_MASK = ~(PAGE_SIZE - 1);
+
+	buf[4095] = '\0';
+	SAFE_FILE_SCANF("/proc/cmdline", "%4095[^\n]", buf);
+
+	if ((p = strstr(buf, "stack_guard_gap=")) != NULL) {
+		if (sscanf(p, "stack_guard_gap=%ld", &GAP_PAGES) != 1) {
+			tst_brk(TBROK | TERRNO, "sscanf");
+			return;
+		}
+		tst_res(TINFO, "stack_guard_gap = %ld", GAP_PAGES);
+	}
+
+	THRESHOLD = (GAP_PAGES - 1) * PAGE_SIZE;
```

Comment:
I can get it to fail with "stack_guard_gap = 1" if I play with stack limit/usage a bit:
```
--- a/testcases/cve/stack_clash.c
+++ b/testcases/cve/stack_clash.c
@@ -133,6 +133,7 @@ void do_child(void)
        void *map;
        unsigned long gap = GAP_PAGES * PAGE_SIZE;
        struct rlimit rlimit;
+       volatile char garbage[990*1024] = "a";
 
        stack_addr = read_stack_addr_from_proc(&stack_size);
        if (stack_addr == -1UL) {

# uname -r
4.12.3
# ulimit -s 1024
# ./stack_clash
tst_test.c:898: INFO: Timeout per run is 0h 05m 00s
stack_clash.c:203: INFO: stack_guard_gap = 1
stack_clash.c:215: INFO: STACK_GROWSDOWN = 1 == 0x3ffc31fd518 > 0x3ffc31fd490
stack_clash.c:161: INFO: Stack:0x3ffc3106000+0xfa000 mmap:0x3ffc3104000+0x1000
stack_clash.c:85: INFO: mmap = [3ffc3104000, 3ffc3105000), addr = 3ffc3103000, diff = ffffffffffffe000, THRESHOLD = 0
stack_clash.c:237: FAIL: stack is too close to the mmaped area
```